### PR TITLE
Codecs (part 14): abi-based encoding & decoding

### PR DESCRIPTION
**Pull request by andreibancioiu**
This is part of a series of pull requests.

See: https://github.com/multiversx/mx-sdk-py/pull/32.

Changes:

 - Define the `Abi` component, able to encode & decode endpoint parameters. When the component is initialized, it reads the ABI definitions and prepares value prototypes (typed values) for endpoint parameters & custom types. Then, when appropriate, it clones these prototypes and populates them with native Python values / gets Python values out of them.
 - Add missing tests for tuples etc.

### CODE REVIEW

File: multiversx_sdk/abi/tuple_value.py [hunk 1]
```diff
@@ -30,3 +30,22 @@ def decode_top_level(self, data: bytes):
         reader = io.BytesIO(data)
         self.decode_nested(reader)
 
+    def set_payload(self, value: Any):
```
Comment (andreibancioiu) [inline multiversx_sdk/abi/tuple_value.py:33]: Forgot about this in previous PRs.

File: multiversx_sdk/abi/tuple_value_test.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import re
+
+import pytest
+
+from multiversx_sdk.abi.small_int_values import U32Value
+from multiversx_sdk.abi.string_value import StringValue
+from multiversx_sdk.abi.tuple_value import TupleValue
+
+
+def test_set_payload_and_get_payload():
```
Comment (andreibancioiu) [inline multiversx_sdk/abi/tuple_value_test.py:10]: Forgot about tuples in the previous PRs.

File: multiversx_sdk/abi/abi_test.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                    )
+                ]
+            )
+        ]
+    )
+
+    assert encoded_values == expected_encoded_values
+
+
+def test_decode_endpoint_output_parameters_real_world_multisig_get_pending_action_full_info():
```
Comment (popenta) [inline multiversx_sdk/abi/abi_test.py:225]: could have skipped `_real_world_` from the name here and above, but not a big deal :)

**Review by popenta — DISMISSED** [review]

File: multiversx_sdk/abi/abi_test.py (hunk 3 above)
Comment (andreibancioiu) [reply multiversx_sdk/abi/abi_test.py]: Indeed, removed (remained in the Go test, for the moment).

https://github.com/multiversx/mx-sdk-abi-go/blob/main/abi/serializer_test.go

**Review by axenteoctavian — APPROVED** [review]

**Review by popenta — APPROVED** [review]

**Pull request merged**